[Feature][Connector-Paimon] Support dynamic bucket splitting improves Paimon writing efficiency

hawk9821 opened the pull request:
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist

  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/seatunnel/issues).

  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.

  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.

-->

### Purpose of this pull request
Support dynamic bucket splitting improves Paimon writing efficiency
<!-- Describe the purpose of this pull request. For example: This pull request adds checkstyle plugin.-->


### Does this PR introduce _any_ user-facing change?
no
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released SeaTunnel versions or within the unreleased branches such as dev.
If no, write 'No'.
If you are adding/modifying connector documents, please follow our new specifications: https://github.com/apache/seatunnel/issues/4544.
-->


### How was this patch tested?
e2e: PaimonSinkDynamicBucketIT 
UT: PaimonBucketAssignerTest#bucketAssigner
e2e case: **PaimonSinkDynamicBucketIT#testPaimonBucketCountOnSparkAndFlink** ，because spark and Flink engine can not auto create paimon table on worker node in local file,  this e2e case work on local hdfs environment.  
![17266244003976](https://github.com/user-attachments/assets/47cd0c04-53f7-4445-bca6-00b6ea97ae79)



<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If you are adding E2E test cases, maybe refer to https://github.com/apache/seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/resources/mysqlcdc_to_mysql.conf, here is a good example.
-->


### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

Hisoka-X commented:
cc @dailai and @TaoZex

Hisoka-X commented on docs/zh/connector-v2/sink/Paimon.md:
```diff
@@ -30,21 +30,20 @@ libfb303-xxx.jar
-| schema_save_mode            | 枚举  | 否    | CREATE_SCHEMA_WHEN_NOT_EXIST | Schema保存模式                                                                                        |
-| data_save_mode              | 枚举  | 否    | APPEND_DATA                  | 数据保存模式                                                                                            |
-| paimon.table.primary-keys   | 字符串 | 否    | -                            | 主键字段列表，联合主键使用逗号分隔(注意：分区字段需要包含在主键字段中)                                                              |
-| paimon.table.partition-keys | 字符串 | 否    | -                            | 分区字段列表，多字段使用逗号分隔                                                                                  |
```
Why deleted this config?

dailai commented on seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-13/src/main/java/org/apache/seatunnel/translation/flink/sink/FlinkSinkWriterContext.java:
```diff
@@ -50,6 +50,11 @@ public int getIndexOfSubtask() {
     }
 
+    @Override
+    public int getNumberOfParallelSubtasks() {
```
Please do not return 0 directly, if this is not necessary to implement, please give a default implementation in the interface, do not implement it here.

dailai commented on seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/sink/FlinkSinkWriterContext.java:
```diff
@@ -47,6 +47,11 @@ public int getIndexOfSubtask() {
     }
 
+    @Override
+    public int getNumberOfParallelSubtasks() {
```
ditto

wuchunfu committed:
[Improve] Update snapshot version to 2.3.8 (#7435)

* [Improve] Update snapshot version to 2.3.8

* [Improve] Update snapshot version to 2.3.8

dailai commented:
Please retrigger the ci.

dailai commented on seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/sink/bucket/PaimonBucketAssigner.java:
```diff
@@ -0,0 +1,85 @@
+
+import java.io.IOException;
+
+public class PaimonBucketAssigner {
```
Please add UT case for this class to vertify the logical.

dailai commented:
Thinks @hawk9821 . Good job. I think your e2e case needs to be added to the case of multi-parallelism, the current case is all single parallelism. In this way, we can effectively verify whether the dynamic bucketing will change depending on the degree of parallelism of the job. Also, I think you should check the bucket count in every case instead of making a separate case. In addition, each of your cases should verify that the dynamic-bucket.target-row-num argument works as expected.

dailai commented:
Why are there so many file changes. Maybe you're having some problems with your git operations. Please reopen a pr which must only has your commits. Then you can link this pr in new pr and close this pr.

hawk9821 commented:
> Why are there so many file changes. Maybe you're having some problems with your git operations. Please reopen a pr which must only has your commits. Then you can link this pr in new pr and close this pr.

Problems caused by rebase ,  resolved

hawk9821 commented:
> Thinks @hawk9821 . Good job. I think your e2e case needs to be added to the case of multi-parallelism, the current case is all single parallelism. In this way, we can effectively verify whether the dynamic bucketing will change depending on the degree of parallelism of the job. Also, I think you should check the bucket count in every case instead of making a separate case. In addition, each of your cases should verify that the dynamic-bucket.target-row-num argument works as expected.

get, add e2e case  **PaimonSinkDynamicBucketIT.testParallelismBucketCount**  **PaimonSinkDynamicBucketIT.testCDCParallelismBucketCount**

dailai commented:
LGTM @Hisoka-X PTAL

Hisoka-X commented on docs/en/connector-v2/sink/Paimon.md:
```diff
@@ -43,7 +43,7 @@ libfb303-xxx.jar
 | paimon.table.primary-keys   | String | No       | -                            | Default comma-separated list of columns (primary key) that identify a row in tables.(Notice: The partition field needs to be included in the primary key fields) |
 | paimon.table.partition-keys | String | No       | -                            | Default comma-separated list of partition fields to use when creating tables.                                                                                    |
-| paimon.table.write-props    | Map    | No       | -                            | Properties passed through to paimon table initialization, [reference](https://paimon.apache.org/docs/master/maintenance/configurations/#coreoptions).            |
+| paimon.table.write-props    | Map    | No       | -                            | Properties passed through to paimon table initialization, [reference](https://paimon.apache.org/docs/0.8/maintenance/configurations/#coreoptions).               |
```
Please revert this change.

Hisoka-X commented on docs/en/connector-v2/sink/Paimon.md:
```diff
@@ -241,6 +241,88 @@ sink {
 
+### Single dynamic bucket table with write props of paimon，operates on the primary key table and bucket is -1
+
+#### core options：[reference](https://paimon.apache.org/docs/0.8/primary-key-table/data-distribution/#dynamic-bucket)
```
```suggestion
core options：[reference](https://paimon.apache.org/docs/master/primary-key-table/data-distribution/#dynamic-bucket)
```

Hisoka-X commented on docs/en/connector-v2/sink/Paimon.md:
```diff
@@ -241,6 +241,88 @@ sink {
+#### core options：[reference](https://paimon.apache.org/docs/0.8/primary-key-table/data-distribution/#dynamic-bucket)
+
+|              name              | type | required | default values |                  Description                   |
+|--------------------------------|------|----------|----------------|------------------------------------------------|
+| dynamic-bucket.target-row-num  | long | 是        | 2000000L       | controls the target row number for one bucket. |
+| dynamic-bucket.initial-buckets | int  | 否        |                | controls the number of initialized bucket.     |
```
Change chinese to english.

Hisoka-X commented on docs/en/connector-v2/sink/Paimon.md:
````diff
@@ -241,6 +241,88 @@ sink {
 }
 ```
 
+### Single dynamic bucket table with write props of paimon，operates on the primary key table and bucket is -1
````
```suggestion
### Write to dynamic bucket table
Single dynamic bucket table with write props of paimon，operates on the primary key table and bucket is -1
```

Hisoka-X replied:
Also please add comment for this feature only worked on zeta.

Hisoka-X commented on docs/en/connector-v2/sink/Paimon.md:
````diff
@@ -241,6 +241,88 @@ sink {
+}
+```
+
+### Single dynamic bucket table with write props of paimon，operates on the primary key table and bucket is -1
+
+#### core options：[reference](https://paimon.apache.org/docs/0.8/primary-key-table/data-distribution/#dynamic-bucket)
+
+|              name              | type | required | default values |                  Description                   |
+|--------------------------------|------|----------|----------------|------------------------------------------------|
+| dynamic-bucket.target-row-num  | long | 是        | 2000000L       | controls the target row number for one bucket. |
+| dynamic-bucket.initial-buckets | int  | 否        |                | controls the number of initialized bucket.     |
+
+```hocon
+env {
+  parallelism = 1
+  job.mode = "STREAMING"
+  checkpoint.interval = 5000
+}
+
+source {
+  Mysql-CDC {
+    base-url = "jdbc:mysql://127.0.0.1:3306/seatunnel"
+    username = "root"
+    password = "******"
+    table-names = ["seatunnel.role"]
+  }
+}
+
+sink {
+  Paimon {
+    catalog_name="seatunnel_test"
+    warehouse="file:///tmp/seatunnel/paimon/hadoop-sink/"
+    database="seatunnel"
````
It's duplicated?

Hisoka-X commented on seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/java/org/apache/seatunnel/e2e/connector/paimon/PaimonSinkDynamicBucketIT.java:
```diff
@@ -0,0 +1,358 @@
+        value = {},
+        type = {EngineType.SPARK, EngineType.FLINK},
+        disabledReason =
+                "Spark and Flink engine can not auto create paimon table on worker node in local file(e.g flink tm) by savemode feature which can lead error")
```
Could you share more detail info about the reason of this error?

Hisoka-X commented on seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/SinkWriter.java:
```diff
@@ -92,6 +92,11 @@ interface Context extends Serializable {
         int getIndexOfSubtask();
 
+        /** @return parallelism of this writer. */
+        default int getNumberOfParallelSubtasks() {
```
I am looking for a way to get spark/flink to support both returning parallelism.

hawk9821 committed:
[Feature][CONNECTORS-V2-Paimon] Dynamic bucket splitting improves Paimon writing efficiency

hawk9821 committed:
[Feature][CONNECTORS-V2-Paimon] spark task parallelism

[Feature][CONNECTORS-V2-Paimon] spark task parallelism

hawk9821 committed:
[Feature][CONNECTORS-V2-Paimon] flink task parallelism

[Feature][CONNECTORS-V2-Paimon] update doc

[Feature][CONNECTORS-V2-Paimon] write to dynamic bucket table , spark flink e2e

Hisoka-X commented on docs/en/connector-v2/sink/Paimon.md:
```diff
@@ -241,6 +241,49 @@ sink {
+
+Single dynamic bucket table with write props of paimon，operates on the primary key table and bucket is -1, only worked on zeta.
+
+#### core options：[reference](https://paimon.apache.org/docs/master/primary-key-table/data-distribution/#dynamic-bucket)
```
```suggestion
#### core options
Please [reference](https://paimon.apache.org/docs/master/primary-key-table/data-distribution/#dynamic-bucket)
```

hawk9821 committed:
[Feature][CONNECTORS-V2-Paimon] Dynamic bucket splitting improves Paimon writing efficiency

Hisoka-X reviewed (approved):
LGTM if ci passes. Thanks @hawk9821
